src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_StringConcat.cs

```diff
@@ -383,7 +383,17 @@ private BoundExpression ConvertConcatExprToStringIfPossible(CSharpSyntaxNode syn
                         if (ConcatExprCanBeOptimizedWithToString(operand.Type))
                         {
                             var toString = GetSpecialTypeMethod(syntax, SpecialMember.System_Object__ToString);
-                            return BoundCall.Synthesized(syntax, operand, toString);
+
+                            var type = (NamedTypeSymbol)operand.Type;
+                            var toStringMembers = type.GetMembers(toString.Name);
```

Comment:
GetMembersUnordered does not filter by name. I think GetMembers(string) are not ordered though.